### ISSUE TITLE
[question] Running karma tests programmatically using open-wc setup

### ISSUE THREAD

**Issue by jarrodek** [open-wc/open-wc#631]
I have a use case (in custom CI pipeline) to run karma tests for my components. I imagined that if I run karma as it is described in [their docs](http://karma-runner.github.io/4.0/dev/public-api.html) and by providing configuration from component's `karma.conf.js` I will be able to run the tests as just running `karma start`. Sadly this is not the case.

The output I see is close to this:

```
START:
23 07 2019 23:21:27.677:WARN [filelist]: Pattern "/tmp/tmp-82898PM7NbYRZPf4/star-rating/__snapshots__/**/*.md" does not match any file.
23 07 2019 23:21:27.775:ERROR [preprocessor.esm]: 

/tmp/tmp-82898PM7NbYRZPf4/star-rating/test/star-rating.test.js: Could not resolve '@open-wc/testing'.
> 1 | import { fixture, assert, aTimeout } from '@open-wc/testing';
    | ^
  2 | import { a11ySuite } from '@advanced-rest-client/a11y-suite/index.js';
  3 | import sinon from 'sinon/pkg/sinon-esm.js';
  4 | import { click, keyDownOn } from '@polymer/iron-test-helpers/mock-interactions.js';
 at /tmp/tmp-82898PM7NbYRZPf4/star-rating/test/star-rating.test.js


23 07 2019 23:21:27.784:INFO [karma-server]: Karma v4.2.0 server started at http://0.0.0.0:9876/
23 07 2019 23:21:27.784:INFO [launcher]: Launching browsers ChromeHeadlessNoSandbox with concurrency unlimited
23 07 2019 23:21:27.787:INFO [launcher]: Starting browser ChromeHeadless
23 07 2019 23:21:28.098:INFO [HeadlessChrome 77.0.3854 (Linux 0.0.0)]: Connected on socket 47zMS4O-i-Ele7ZpAAAA with id 32221475

....

Uncaught TypeError: Failed to resolve module specifier "@open-wc/testing". Relative references must start with either "/", "./", or "../".
at http://localhost:9876/context.html:0:0
```

The dependencies are installed correctly and I can read the files from `node_models` directory.

I understand that the right module is not executed. So the question is how to run the test programmatically? So far I came up with the following script:

```javascript
// Base class downloads git repo, installs dependencies 
// and calls `_run()`. Nothing related to test itself.
class KarmaTestRunner extends BaseTestRunner {
  async _run() {
    const opts = this.createConfig();
    this.server = new Server(opts, (exitCode) => {
      logging.verbose(`Karma has exited with ${exitCode}`);
    });
    this._addListeners(this.server);
    this.server.start();
  }

  createConfig() {
    // this.componentDir is where temporary files were cloned to
    return config.parseConfig(path.join(this.componentDir, 'karma.conf.js'), {
      basePath: this.componentDir
    } );
  }

  _addListeners(srv) {
    srv.on('browser_complete', this._browserComplete.bind(this));
    srv.on('browser_error', this._browserError.bind(this));
    srv.on('run_complete', this._runComplete.bind(this));
  }

  _browserComplete(browser, result) {
    console.log('_browserComplete');
    console.log(browser, result);
  }

  _browserError(browser, error) {
    console.log('_browserError');
    console.log(browser, error);
  }

  _runComplete(browsers, results) {
    console.log('_runComplete');
    console.log(browsers, results);
  }
}
``` 

The configuration object looks correct. It has configuration read from component's `karma.conf.js` file and paths resolved to the tmp location. However it still fails. This is how the final config object looks like:

```
Config {
  LOG_DISABLE: 'OFF',
  LOG_ERROR: 'ERROR',
  LOG_WARN: 'WARN',
  LOG_INFO: 'INFO',
  LOG_DEBUG: 'DEBUG',
  frameworks:
   [ 'mocha',
     'snapshot',
     'mocha-snapshot',
     'source-map-support',
     'esm' ],
  protocol: 'http:',
  port: 9876,
  listenAddress: '0.0.0.0',
  hostname: 'localhost',
  httpsServerConfig: {},
  basePath: '/tmp/tmp-82898PM7NbYRZPf4/star-rating',
  files:
   [ Pattern {
       pattern:
        '/tmp/tmp-82898PM7NbYRZPf4/star-rating/__snapshots__/**/*.md',
       served: true,
       included: true,
       watched: true,
       nocache: false,
       weight: [Array],
       type: undefined },
     Pattern {
       pattern:
        '/tmp/tmp-82898PM7NbYRZPf4/star-rating/node_modules/axe-core/axe.min.js',
       served: true,
       included: true,
       watched: true,
       nocache: false,
       weight: [Array],
       type: undefined },
     Pattern {
       pattern: '/tmp/tmp-82898PM7NbYRZPf4/star-rating/test/**/*.test.js',
       served: true,
       included: true,
       watched: true,
       nocache: false,
       weight: [Array],
       type: 'module' } ],
  browserConsoleLogOptions: { level: 'debug', format: '%b %T: %m', terminal: true },
  customContextFile: null,
  customDebugFile: null,
  customClientContextFile: null,
  exclude: [ '/tmp/tmp-82898PM7NbYRZPf4/star-rating/karma.conf.js' ],
  logLevel: 'INFO',
  colors: true,
  autoWatch: false,
  autoWatchBatchDelay: 250,
  restartOnFileChange: true,
  usePolling: true,
  reporters: [ 'mocha' ],
  singleRun: true,
  browsers: [ 'ChromeHeadlessNoSandbox' ],
  captureTimeout: 60000,
  proxies: {},
  proxyValidateSSL: true,
  preprocessors:
   [Object: null prototype] {
     '/tmp/tmp-82898PM7NbYRZPf4/star-rating/**/__snapshots__/**/*.md': [ 'snapshot', 'snapshot-filename' ],
     '/tmp/tmp-82898PM7NbYRZPf4/star-rating/**/*.test.js': [ 'esm' ],
     '/tmp/tmp-82898PM7NbYRZPf4/star-rating/**/*.spec.js': [ 'esm' ] },
  urlRoot: '/',
  upstreamProxy: undefined,
  reportSlowerThan: 0,
  loggers: [ { type: 'console', layout: [Object] } ],
  transports: [ 'polling', 'websocket' ],
  forceJSONP: false,
  plugins:
   [ '/tmp/tmp-82898PM7NbYRZPf4/star-rating/node_modules/karma-mocha/lib/index.js',
     '/tmp/tmp-82898PM7NbYRZPf4/star-rating/node_modules/karma-mocha-reporter/index.js',
     '/tmp/tmp-82898PM7NbYRZPf4/star-rating/node_modules/karma-source-map-support/lib/index.js',
     '/tmp/tmp-82898PM7NbYRZPf4/star-rating/node_modules/karma-coverage-istanbul-reporter/src/reporter.js',
     '/tmp/tmp-82898PM7NbYRZPf4/star-rating/node_modules/karma-static/index.js',
     '/tmp/tmp-82898PM7NbYRZPf4/star-rating/node_modules/karma-snapshot/lib/index.js',
     '/tmp/tmp-82898PM7NbYRZPf4/star-rating/node_modules/karma-mocha-snapshot/lib/index.js',
     '/tmp/tmp-82898PM7NbYRZPf4/star-rating/node_modules/karma-chrome-launcher/index.js',
     '/tmp/tmp-82898PM7NbYRZPf4/star-rating/node_modules/@open-wc/testing-karma/src/snapshot-filename-preprocessor.js',
     'karma-*',
     '/tmp/tmp-82898PM7NbYRZPf4/star-rating/node_modules/@open-wc/karma-esm/src/index.js',
     [Object: null prototype] { 'launcher:ChromeHeadlessNoSandbox': [Array] } ],
  client:
   { args: [],
     useIframe: true,
     runInParent: false,
     captureConsole: true,
     clearContext: true,
     mocha: { reporter: 'html' } },
  defaultClient:
   { args: [],
     useIframe: true,
     runInParent: false,
     captureConsole: true,
     clearContext: true,
     mocha: { reporter: 'html' } },
  browserDisconnectTimeout: 2000,
  browserDisconnectTolerance: 0,
  browserNoActivityTimeout: 30000,
  processKillTimeout: 2000,
  concurrency: Infinity,
  failOnEmptyTestSuite: true,
  retryLimit: 2,
  detached: false,
  crossOriginAttribute: true,
  browserSocketTimeout: 20000,
  customLaunchers:
   { ChromeHeadlessNoSandbox: { base: 'ChromeHeadless', flags: [Array] } },
  middleware: [ 'static', 'esm' ],
  static:
   { path:
      '/home/pawel/workspace/advanced-rest-client/api-components-apps/ci-app' },
  mochaReporter: { showDiff: true },
  formatError: [Function: formatError],
  coverageIstanbulReporter:
   { reports: [ 'html', 'lcovonly', 'text-summary' ],
     dir: 'coverage',
     combineBrowserReports: true,
     skipFilesWithNoCoverage: false,
     thresholds: { global: [Object] } },
  snapshot:
   { update: undefined,
     prune: undefined,
     limitUnusedSnapshotsInWarning: -1,
     pathResolver: [Function: pathResolver] },
  esm: { coverage: undefined, babel: { exclude: [Array] } } }
```

When I run test command in the directory it works as expected:

```
pawel@pawel-pc-ubuntu:/tmp/tmp-9049xD899ZGWHntL/star-rating$ ./node_modules/.bin/karma start
```

Any idea how can I make this work?

**Comment by LarsDenBakker** [comment]
Sounds like an issue with your CI or karma. Are you sure it's installing your dependencies? You could temporarily try to use the absolute path to `@open-wc/testing` to rule out there isn't any issue with the module resolving.

**Comment by jarrodek** [comment]
I've checked if paths exists and all of them are absolute and correct.
What intrigues me is this fragment of the configuration:

```
static:
   { path:
      '/home/pawel/workspace/advanced-rest-client/api-components-apps/ci-app' },
```
This is not right as this is from where the scrip is executed. It might be an issue with karma but I am not sure.

I tried something else instead. Before reading configuration and running the server I change script directory to component's location and then is is actually working. I was hoping that the `basePath` in configuration do the trick. I think this is on karma. Because I made this work and apparently it is not `open-wc`'s issues I am closing this.
